### PR TITLE
Added logic for dropping LiDAR offshore

### PR DESCRIPTION
**Pull request by rosepearson**
Add logic to the GeoFabricsGenerator so that all LiDAR information away from the foreshore is discarded.
- [X] Make code change
- [X] Update tests to include 'drop_lidar_offshore' in their instruction files
  - [x] Change one test to drop offshore LiDAR
  - [x] Fix the local files test to work correctly again - had to drop out the reference DEM
- [x] Ensure passing again on GH actions
- [x] Test locally on the Westport and Wellington regions for expected behaviour
- [x] Update documentation

### CODE REVIEW

**Comment by rosepearson** [comment]
Hi @AliceHarang - another PR - obviously don't look at it until you're back at work Monday :) This is a minor change to address the fact I was using LiDAR that was the ocean as you pointed out in the Westport DEM.

**Comment by rosepearson** [comment]
@AliceHarang I've finished with my changes. I am currently running this over the Westport region to generate an updated DEM. I have already run it over the Waikanae region is testing. 
As part of this I have updated the YML files to use more up-to-date libraries. If you choose to test locally on your machine you will probably need to update your environment first.

File: src/geofabrics/processor.py [hunk 1]
```diff
@@ -301,15 +306,17 @@ def run(self):
...
             # update the dense DEM with a patch created from the LiDAR tile
-            self.dense_dem.add_tile(self.catchment_lidar.tile_array, window_size, idw_power, radius)
+            self.dense_dem.add_tile(tile_points=self.catchment_lidar.tile_array,
+                                    tile_extent=self.catchment_lidar.tile_extent,
+                                    window_size=window_size, idw_power=idw_power, radius=radius)
             del self.catchment_lidar.tile_array
 
         # Filter the LiDAR extents based on the area_to_drop
-        self.catchment_lidar.filter_lidar_extents_for_holes()
+        self.dense_dem.filter_lidar_extents_for_holes()
```
Comment (AliceHarang) [inline src/geofabrics/processor.py:315]: Can you clarify in comment what is a "dense_dem"?

File: tests/test_processor_remote_tiles_westport/instruction.json [hunk 2]
```diff
@@ -18,10 +18,11 @@
 			"benchmark_dem": "tests/test_processor_remote_tiles_westport/data/benchmark_dem.nc"
 		},
 		"apis": {
-			"open_topography": "True"
+			"open_topography": true
 		},
 		"general": {
-			"set_dem_shoreline": "True"
+			"set_dem_shoreline": true,
+			"drop_offshore_lidar": false
```
Comment (AliceHarang) [inline tests/test_processor_remote_tiles_westport/instruction.json:25]: did you test with "droping the offshore lidar" for all your test cases?

File: src/geofabrics/dem.py [hunk 3]
```diff
@@ -104,14 +105,14 @@ def _extract_points(self):
...
+    def points(self) -> numpy.ndarray:
         """ The reference DEM points after any extent or foreshore value
         filtering. """
 
         return self._points
 
     @property
-    def extents(self):
+    def extents(self) -> geopandas.GeoDataFrame:
         """ The extents for the reference DEM """
```
Comment (AliceHarang) [inline src/geofabrics/dem.py:116]: Can you clarify this term in comments or readme somewhere? (If it is not already done  :-) )
What type of info it is? (cache matrix or contours? holes included or just regarding boundaries? )
It is for each tile or for the full domain?

File: src/geofabrics/dem.py [hunk 4]
```diff
@@ -241,13 +244,71 @@ def add_tile(self, tile_points: numpy.ndarray, window_size: int, idw_power: int,
...
+                self._extents = geopandas.GeoDataFrame(
+                    {'geometry': [shapely.ops.cascaded_union([self._extents.loc[0].geometry,
+                                                              tile_extent.loc[0].geometry])]},
+                    crs=self.catchment_geometry.crs['horizontal'])
+            self._extents = geopandas.clip(self.catchment_geometry.catchment, self._extents)
+
+    def filter_lidar_extents_for_holes(self):
+        """ Remove holes below a filter size within the extents """
+
+        if self.area_to_drop is None:
```
Comment (AliceHarang) [inline src/geofabrics/dem.py:274]: Do you still check/remove small holes if you have a demand to drop area offshore?

File: src/geofabrics/processor.py (hunk 1 above)
Comment (rosepearson) [reply src/geofabrics/processor.py:315]: Good idea. Will do :)

File: tests/test_processor_remote_tiles_westport/instruction.json (hunk 2 above)
Comment (rosepearson) [reply tests/test_processor_remote_tiles_westport/instruction.json:25]: Three tests have is set to False, and one test has it set to True. That test is the 'test_processor_remote_all_westport'

**Review by AliceHarang — APPROVED** [review]
Hi Rose, 
I am happy for you to merge this PR when you think you have answer enough of my questions.
Cheers

PS: if somes are inappropriate or miss sth, let me know!

**Comment by rosepearson** [comment]
> Hi Rose,
> I am happy for you to merge this PR when you think you have answer enough of my questions.
> Cheers
> 
> PS: if somes are inappropriate or miss sth, let me know!

Thanks Alice. Comments are great as always. Will work through them now :)

File: src/geofabrics/processor.py (hunk 1 above)
Comment (rosepearson) [reply src/geofabrics/processor.py:316]: I've added a description of the dense_dem object and all others (catchment_lidar, etc) in the `GeoFabricsGenerator` doc-string.

File: src/geofabrics/dem.py (hunk 3 above)
Comment (rosepearson) [reply src/geofabrics/dem.py:116]: Good idea. Adding a description now

File: src/geofabrics/dem.py (hunk 4 above)
Comment (rosepearson) [reply src/geofabrics/dem.py:282]: Yup :) - added a comment to make this clearer!

File: src/geofabrics/dem.py (hunk 3 above)
Comment (rosepearson) [reply src/geofabrics/dem.py:116]: Added and also in the wiki. Hopefully clearer now.

**Pull request merged**